### PR TITLE
Polish agent and team sharing and snapshots

### PR DESCRIPTION
**Pull request by klopez4212**
## Summary
- replace catalog and snapshot menu actions with a focused Share flow for people, links, memory scope, and export
- extend the same sharing and export experience to agent teams, including direct sharing, copied links, and composer attachment cards
- simplify PNG-first export controls and animate memory-warning layout changes
- render shared agents and teams as native attachments with clear names, attribution, download, and Add actions

## Verification
- pnpm check
- pnpm test
- desktop Tauri tests
- pnpm run typecheck
- pnpm run build
- focused Playwright coverage for agent and team sharing, export, copied-link cards, recipient attachments, and picker interactions

### Share agent
Invite people, choose the agent and memory scope, or copy a workspace link. Teams now use the same sharing treatment.

Share agent

### Export agent
Choose memory scope and PNG or JSON before exporting. The same controls are available for teams.

Export agent

### Shared agent attachment
Received agents and teams use the native attachment treatment with Download and Add actions.

Shared agent attachment

### CODE REVIEW

File: desktop/src/features/agents/ui/PersonaShareDialog.tsx [hunk 1]
```diff
@@ -10,96 +26,319 @@ import {
...
+
+  async function uploadSnapshot(
+    memoryLevel: SnapshotMemoryLevel,
+  ): Promise<BlobDescriptor> {
+    const encoded = await encodeSnapshotMutation.mutateAsync({
+      id: persona.id,
+      memoryLevel: hasLinkedAgent ? memoryLevel : "none",
+      format: "png",
+      memorySourcePubkey: linkedAgentPubkey,
+      avatarPngDataUrl: await resolveSnapshotAvatarPng(persona.avatarUrl),
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/agents/ui/PersonaShareDialog.tsx]: 🤖 Fixed in 7b2c989f7. Memory-bearing Copy link and Send actions now require an explicit Share memories? confirmation before encoding or uploading. The confirmation identifies the plaintext memory scope and link exposure, and the E2E coverage verifies that encoding does not start until confirmation.

File: desktop/src/features/agents/ui/usePersonaActions.ts [hunk 2]
```diff
@@ -128,9 +115,9 @@ export function usePersonaActions() {
   const [snapshotImportConfirmError, setSnapshotImportConfirmError] =
     React.useState<string | null>(null);
   const [isCatalogDialogOpen, setIsCatalogDialogOpen] = React.useState(false);
-  const [sharedCatalogPersonaIds, setSharedCatalogPersonaIds] = React.useState<
-    string[]
-  >(readSharedCatalogPersonaIds);
+  const [sharedCatalogPersonaIds] = React.useState<string[]>(
+    readSharedCatalogPersonaIds,
+  );
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/agents/ui/usePersonaActions.ts]: 🤖 Fixed in 7b2c989f7. The retired catalog visibility IDs no longer feed catalog state, and the agents view now removes the legacy buzz-persona-catalog-visibility-v1 preference. Unit coverage verifies the cleanup and its storage-error fallback.

File: desktop/src/features/agents/ui/PersonaShareDialog.tsx [hunk 3]
```diff
@@ -10,96 +26,319 @@ import {
...
+    try {
+      const uploaded = await uploadSnapshot(linkShareLevel);
+      await navigator.clipboard.writeText(uploaded.url);
+      toast.success("Link copied");
+    } catch {
+      toast.error("Couldn’t copy link");
+    } finally {
+      setIsCopying(false);
+    }
+  }
```
Comment (wpfleger96) [inline desktop/src/features/agents/ui/PersonaShareDialog.tsx]: **[CRITICAL — blocking]**

With `Agent + core memory` or `Agent + all memories` selected, clicking either `Copy link` or `Send` immediately encodes and uploads plaintext agent memory. The old flow gated this behind an explicit confirmation that named the destination and warned that memory travels in plaintext and that anyone holding the media link can fetch it — that gate didn't make it into this dialog (the Export dialog kept its warning; this path has none).

Separately, the footer says "Anyone in this workspace with the link can duplicate and use this agent" — but media GET isn't workspace-authenticated. `get_blob` / `bind_media_read_tenant` (`crates/buzz-relay/src/api/media.rs:513`, `:421-432`) only bind Host→tenant and never check who the requester is; possession of the URL is sufficient, workspace member or not. Since `handleCopyLink()` puts that raw URL on the clipboard, the copy promises an access control that doesn't exist — a rough combination with plaintext memory in the payload.

Could we (a) restore a confirmation step before upload whenever a memory level is selected — naming the recipients or link audience, and stating that anyone with the URL can retrieve the bytes — and (b) fix the footer copy so it doesn't claim workspace scoping?

File: desktop/src/features/agents/ui/PersonaShareDialog.tsx [hunk 4]
```diff
@@ -10,96 +26,319 @@ import {
...
+        outgoingMessage.mediaTags ?? [],
+      );
+      toast.success(`Sent ${persona.displayName}`);
+      onOpenChange(false);
+    } catch {
+      toast.error("Couldn’t send agent. Try again.");
+    } finally {
+      setIsSending(false);
+    }
+  }
```
Comment (wpfleger96) [inline desktop/src/features/agents/ui/PersonaShareDialog.tsx]: **[CRITICAL — blocking]**

This send path skips the guarded pipeline the old flow ran everything through. `handleSend()` goes straight `openDm` → upload → `sendChannelMessage`, while the old controller enforced: the timeout action-boundary check, moderation-DM exclusion that fails closed while identity/relay-self are unresolved, archived/non-member/forum destination blocking, a double-send guard, and eligibility rechecks both before encode and again before upload (`checkSendEligibility`, `useSnapshotSendController.ts:135`). None of that runs here, and the recipient picker only filters self/agents/already-selected/archived (`PersonaShareRecipients.tsx:60-72`) — no relay-moderation-identity exclusion, no timeout check. Net effect: a timed-out user can encode and upload memory before anything rejects the send, and the moderation identity can be selected from directory search.

The controller was built to be reusable — I'd route this through it (or extract the policy into a recipient-based send service both paths share), keeping eligibility checked before encode and again before upload, and add tests pinning those boundaries on this path since the old E2Es covering them were deleted.

File: desktop/src/shared/ui/markdownFileCard.ts [hunk 5]
```diff
@@ -40,6 +42,23 @@ export type ResolvedSnapshotCard = {
...
+    .replace(/\.agent\.(?:json|png)$/i, "")
+    .replace(/[-_]+/g, " ")
+    .trim();
+  if (!filenameStem) return "Agent";
+
+  return filenameStem
+    .split(/\s+/)
+    .map((word) => `${word.charAt(0).toUpperCase()}${word.slice(1)}`)
+    .join(" ");
+}
```
Comment (wpfleger96) [inline desktop/src/shared/ui/markdownFileCard.ts:60]: **[IMPORTANT — blocking]**

This promotes sender-controlled text to the agent's identity before verification. `snapshotDisplayName()` takes the markdown link label — entirely sender-authored — and makes it the attachment title. The previous card was deliberate about the opposite: filename plus a neutral "Agent snapshot" classification, with a comment noting the sender-supplied agent name is never shown pre-verification (old `AgentSnapshotCard.tsx:42-44`). With this change, `[Payroll Approver](…/x.agent.png)` renders as a trusted-looking native attachment titled "Payroll Approver" before a single byte is verified. (Import itself still holds — the bounded sha256-verified fetch is intact — this is about what the card advertises pre-import.)

I'd keep the pre-verification title neutral/filename-derived and only surface the decoded agent name after the verified fetch — the import preview already does exactly that.

File: desktop/src/features/messages/ui/MessageRow.tsx [hunk 6]
```diff
@@ -347,6 +347,7 @@ export const MessageRow = React.memo(
               mentionNames={mentionNames}
               mentionPubkeysByName={mentionPubkeysByName}
               searchQuery={searchQuery}
+              snapshotSharedBy={message.author}
```
Comment (wpfleger96) [inline desktop/src/features/messages/ui/MessageRow.tsx]: **[IMPORTANT — blocking]** (pairs with the `markdownFileCard.ts` comment)

`message.author` here is the tag-attributed display author, which is spoofable via `actor`/`p` attribution — the comment ten lines up on `configNudgeAuthorPubkey` explains exactly why trust-sensitive claims must use the raw event signer instead. "Shared by X" on an agent snapshot card is a provenance claim, so it should be derived from the signer identity — or dropped from the card.

File: desktop/src/features/agents/ui/AgentSnapshotSendDialog.tsx [hunk 7]
```diff
@@ -170,6 +170,7 @@ export function AgentSnapshotSendDialog({
           avatarPngDataUrl: await resolveSnapshotAvatarPng(persona.avatarUrl),
         }),
       destination.id,
+      persona.displayName,
```
Comment (wpfleger96) [inline desktop/src/features/agents/ui/AgentSnapshotSendDialog.tsx]: **[IMPORTANT — blocking]**

With "Send in Buzz" removed from the menu, nothing in production imports `AgentSnapshotSendDialog` anymore — which strands this dialog (507 lines), `useSnapshotSendController.ts` (503), and `agentSnapshotSend.test.mjs` (791) as unreachable code. Meanwhile the PR deletes all 17 E2E tests in `agent-snapshot-send.spec.ts` that pinned the old safety contract (moderation-DM fail-closed, timeout blocking, destination invalidation before/during encode, memory-disclosure gating, duplicate-send protection). The remaining unit tests now pass against code users can't reach, which reads as coverage we no longer have.

If the guarded controller gets reused by the new share flow (see the `handleSend` comment), most of this resolves itself. Whatever is genuinely dead after that should be deleted in this same PR, and the safety behaviors need equivalent tests on the new active path.

File: desktop/src/features/agents/ui/usePersonaActions.ts (hunk 2 above)
Comment (wpfleger96) [inline desktop/src/features/agents/ui/usePersonaActions.ts]: **[IMPORTANT — blocking]**

The writer for `buzz-persona-catalog-visibility-v1` is deleted but this reader stays, so the localStorage state becomes read-only: anyone who previously shared a persona to the catalog now has it stuck there with no UI to remove it, and nothing new can ever be added. If catalog visibility is being removed as a feature, the reader and legacy key should go too (or be migrated/ignored); if it's staying, it needs a management control somewhere. Half-alive like this is the worst of both.

File: desktop/src/features/agents/ui/PersonaShareDialog.tsx [hunk 8]
```diff
@@ -10,96 +26,319 @@ import {
...
+      setIsCopying(false);
+    }
+  }
+
+  async function handleSend() {
+    if (isActionPending || selectedRecipients.length === 0) return;
+
+    setIsSending(true);
+    try {
+      const directMessage = await openDmMutation.mutateAsync({
```
Comment (wpfleger96) [inline desktop/src/features/agents/ui/PersonaShareDialog.tsx]: **[Question — blocking only if unintentional]**

Capability question: the old destination picker could send a snapshot into any joined channel as well as existing DMs/group DMs. The new flow is people→`open_dm` only, so sharing an agent into a channel is no longer possible from any UI surface. Is dropping channel sharing intentional product direction, or a casualty of the redesign? Totally fine if intentional — I just want it to be a decision rather than an accident.

**Review by wpfleger96 — CHANGES_REQUESTED** [review]
Kenny — the visual direction here is genuinely good. The share dialog, the people picker, and the native attachment card for recipients are all a clear step up, and I'd like to keep the design as-is. But a set of safety behaviors the old send flow deliberately enforced got dropped on the way to the new UI, and a couple of capabilities lost parity, so I'm requesting changes before this lands.

To set scope: renames and reorganizations are all fine — folding Export snapshot into Share, the menu restructure, etc. The asks below are about capability and safety parity, not naming.

Blocking (details inline):
1. Memory-bearing shares upload plaintext agent memory with no confirmation step, and the link footer makes an access claim the relay doesn't actually enforce.
2. `handleSend()` bypasses the guarded send pipeline entirely (timeout, moderation-DM fail-closed, destination eligibility, double-send guard).
3. The recipient card now renders sender-controlled text as the agent's identity before any verification, plus a spoofable "Shared by" attribution.
4. ~1,800 lines of the old send stack are left in-tree unreachable, while the 17 E2E tests that pinned the old safety contract were deleted.
5. Catalog visibility becomes unmanageable zombie state (writer deleted, reader kept).

One capability question, non-blocking if intentional: channel sharing is gone (people→DM only now) — flagged inline.

One scope note for reviewers: this PR doesn't touch team sharing/export at all — team-managed personas gain the new Share item, but it shares a single agent snapshot.

File: desktop/src/shared/ui/markdownFileCard.ts (hunk 5 above)
Comment (wpfleger96) [reply desktop/src/shared/ui/markdownFileCard.ts:60]: actually I changed my mind about this one!

File: desktop/src/features/agents/ui/usePersonaActions.ts (hunk 2 above)
Comment (wpfleger96) [reply desktop/src/features/agents/ui/usePersonaActions.ts]: we can leave this for now, we're gonna remove the catalog for now then re-introduce it after it's reworked

File: desktop/src/features/agents/ui/PersonaShareDialog.tsx (hunk 8 above)
Comment (wpfleger96) [reply desktop/src/features/agents/ui/PersonaShareDialog.tsx]: this is actually fine too - "copy link" covers this since you can just paste the link into what channel you want

File: desktop/src/features/agents/ui/PersonaShareDialog.tsx (hunk 3 above)
Comment (klopez4212) [reply desktop/src/features/agents/ui/PersonaShareDialog.tsx]: 🤖 Addressed in 9c36e36e1. Memory-bearing Copy link and Send actions now require confirmation before encoding or upload; the warning names the selected recipients or link audience, states the plaintext memory scope and URL exposure, and the footer now says “Anyone with the link…” without claiming workspace-only access. The active-share E2E coverage pins the confirmation and corrected footer.

File: desktop/src/features/agents/ui/PersonaShareDialog.tsx (hunk 4 above)
Comment (klopez4212) [reply desktop/src/features/agents/ui/PersonaShareDialog.tsx]: 🤖 Addressed in 9c36e36e1. People sharing now runs through the guarded controller, covering the full action from DM creation through send: same-tick duplicate protection, timeout checks, destination eligibility before and after encoding, and fail-closed moderation-DM classification. The relay moderation identity is also excluded from search, with active-path E2E coverage for each boundary.

File: desktop/src/features/messages/ui/MessageRow.tsx (hunk 6 above)
Comment (klopez4212) [reply desktop/src/features/messages/ui/MessageRow.tsx]: 🤖 Addressed in 9c36e36e1. “Shared by” now resolves from the raw event signer pubkey rather than the actor-attributed display author. Signer pubkeys are included in profile loading, and unit coverage verifies that an actor tag cannot replace the provenance label.

File: desktop/src/features/agents/ui/AgentSnapshotSendDialog.tsx (hunk 7 above)
Comment (klopez4212) [reply desktop/src/features/agents/ui/AgentSnapshotSendDialog.tsx]: 🤖 Addressed in 9c36e36e1. The unreachable AgentSnapshotSendDialog was deleted, the guarded controller is now used by the active Share flow, and the obsolete dialog-only tests were removed. Replacement active-path E2Es cover relay-identity loading and moderation exclusion, timeout blocking, post-encode destination invalidation, and duplicate-send protection.

File: desktop/src/features/agents/ui/PersonaShareDialog.tsx [hunk 9]
```diff
@@ -10,97 +30,461 @@ import {
...
+    setPendingMemoryShare(null);
+    if (pendingShare.action === "copy") {
+      void copyLink(pendingShare.memoryLevel);
+    } else {
+      void sendToRecipients(pendingShare.memoryLevel);
+    }
+  }
 
   return (
     <Dialog onOpenChange={onOpenChange} open={open}>
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/agents/ui/PersonaShareDialog.tsx]: 🤖 Fixed in ed82a137e. Dialog close requests are now ignored while Copy link or Send is pending, so Escape/outside-click cannot unmount and reset the action guard. Added an integration test covering Escape during a delayed send.

File: desktop/src/features/messages/lib/imetaMediaMarkdown.ts [hunk 10]
```diff
@@ -281,6 +287,7 @@ export function buildOutgoingMessage(
   let content = body;
   for (const d of pendingImeta) {
     content += formatImetaMediaLine(d, {
+      label: d.displayLabel,
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/messages/lib/imetaMediaMarkdown.ts:328]: 🤖 Fixed in ed82a137e. Edit mode now restores the durable Markdown attachment label onto the imeta entry before seeding composer attachments, preserving agent names across edits. Added unit coverage for escaped labels, unmatched media, and trailing-label precedence.

File: desktop/src/features/agents/ui/PersonaShareDialog.tsx [hunk 11]
```diff
@@ -10,97 +38,530 @@ import {
...
+  async function sendToRecipients(memoryLevel: SnapshotMemoryLevel) {
+    if (isActionPending || selectedRecipients.length === 0) return;
+
+    const sent = await snapshotSendController.beginSend(
+      () => encodeSnapshot(memoryLevel),
+      async () => {
+        const directMessage = await openDmMutation.mutateAsync({
+          pubkeys: selectedRecipients.map((recipient) => recipient.pubkey),
+        });
+        return directMessage.id;
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/agents/ui/PersonaShareDialog.tsx:404]: 🤖 Fixed in ed82a137e. The Channel returned by open_dm is now reconciled back into the shared channel cache after the active refresh settles, before the send pipeline begins its eligibility checks. Existing stale-refresh reconciliation tests and the focused share E2E flow pass.

File: desktop/src/features/messages/lib/agentSnapshotClipboard.ts [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    !filename.toLowerCase().endsWith(".agent.png") ||
+    !sha256 ||
+    !/^[0-9a-f]{64}$/.test(sha256) ||
+    typeof candidate.size !== "number" ||
+    !Number.isInteger(candidate.size) ||
+    candidate.size <= 0 ||
+    candidate.size > MAX_SNAPSHOT_BYTES ||
+    candidate.type !== "image/png" ||
+    typeof candidate.url !== "string" ||
+    !isSafeSnapshotUrl(candidate.url)
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/messages/lib/agentSnapshotClipboard.ts:135]: 🤖 Fixed in ed82a137e. Clipboard snapshot URLs now reject whitespace and Markdown URL delimiters before becoming composer attachments. Added coverage for closing-parenthesis, newline, and space payloads.

**Review by wesbillman — COMMENTED** [review]
Code pass is clean at `ed82a137e`: the latest hardening addresses the prior send-concurrency, live destination eligibility, attribution, stale-DM, edit-label, pending-close, memory-confirmation, and clipboard-validation findings. CI is green.

I am withholding approval because the head no longer contains current `main` and GitHub now reports the PR as conflicting. The overlap includes agent snapshot/message rendering paths changed by #1846. Please rebase/merge current `main`, resolve the conflicts, and rerun CI; then this is ready for a final approval gate.

**Comment by klopez4212** [comment]
🤖 Update: this PR now applies the sharing work to agent teams as well. Teams use the same Share flow, memory scope and export controls, copied-link and direct-share behavior, and native composer and received-message attachment cards. The old separate team snapshot send dialog has been removed.

File: desktop/src/features/agents/ui/PersonaShareDialog.tsx [hunk 13]
```diff
@@ -11,96 +40,722 @@ import {
...
+  );
+  const isCopying = copyStatus === "copying";
+  const copyStatusLabel =
+    copyStatus === "copying"
+      ? "Copying…"
+      : copyStatus === "copied"
+        ? "Copied"
+        : "Copy link";
+  const isActionPending = isPending || isCopying || isSending;
+  const isInterfacePending = isPending || isSending;
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/agents/ui/PersonaShareDialog.tsx:274]: 🤖 Fixed in 45427b52b. Copy link now participates in the action close guard, so close, Escape, and Export cannot unmount the dialog during upload while the remaining share controls stay visually active. I also added delayed-upload E2E coverage for the guarded state.

File: desktop/src/features/messages/lib/agentSnapshotClipboard.ts [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+    filename.length > MAX_FILENAME_LENGTH ||
+    filename.includes("/") ||
+    filename.includes("\\") ||
+    !/\.(?:agent|team)\.png$/i.test(filename) ||
+    !sha256 ||
+    !/^[0-9a-f]{64}$/.test(sha256) ||
+    typeof candidate.size !== "number" ||
+    !Number.isInteger(candidate.size) ||
+    candidate.size <= 0 ||
+    candidate.size > MAX_SNAPSHOT_BYTES ||
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/messages/lib/agentSnapshotClipboard.ts]: 🤖 Fixed in e20baf400. Clipboard parsing now applies the 10 MiB cap to `.agent.png` and the supported 50 MiB cap to `.team.png`. Added coverage for a valid 11 MiB team snapshot plus oversized agent and team rejection.

File: desktop/src/features/messages/lib/agentSnapshotClipboard.ts [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+    payload = JSON.parse(decodeURIComponent(encodedPayload));
+  } catch {
+    return null;
+  }
+
+  if (!payload || typeof payload !== "object") return null;
+  const candidate = payload as Partial<SnapshotClipboardPayload>;
+  const displayName = candidate.displayName?.trim();
+  const filename = candidate.filename?.trim();
+  const sha256 = candidate.sha256?.trim().toLowerCase();
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/messages/lib/agentSnapshotClipboard.ts]: 🤖 Fixed in 0c7e317d4. Untrusted `displayName`, `filename`, and `sha256` values are now runtime-checked before trimming, so malformed clipboard payloads return `null` and preserve normal paste behavior. Added coverage for non-string values across all three fields.

**Review by wpfleger96 — APPROVED** [review]
